harvard-lts/mirador-url-sync-plugin
# Update canvas query parameter in URL on SET_CANVAS

**JIRA Ticket**: [LTSVIEWER-199](https://jira.huit.harvard.edu/browse/LTSVIEWER-199)

# What does this Pull Request do?

Uses [Redux-saga](https://redux-saga.js.org/)(the current Mirador approved approach) to listen to events from the Mirador redux store, matching "SET_CANVAS", and then updating the URL to contain the correct value in the `canvasId` parameter. 

This works in unison with an update to the MPS-viewer ([see the companion PR](https://github.com/harvard-lts/mps-viewer/pull/47)), which initialises Mirador with an initial `canvasId` taken from the URL.

The acceptance criteria for 199, and this PR, is only concerned:
- The correct `canvasId` appearing in the URL
- The viewer opening at the correct canvas when the page is refreshed or the URL is copy/pasted

# How should this be tested?

**In terms of MPS viewer - this can only be verified by looking at the iframe src page using the instructions below — the URL for the example pages will not update, and is not intended to**
 
- Checkout the [MPS viewer on the branch for the companion PR](https://github.com/harvard-lts/mps-viewer/pull/47)
- Restart your local viewer docker container, and ensure it ran `npm` and `webpack`
- Navigating to one of the examples you should be able to get the `iframe` src in one of a few ways:
   - (Chrome) Right clicking on the sidebar of the viewer you can select `View frame source`, then just delete `view-source:` from the front of the address bar
   - (Firefox) Right clicking on the sidebar of the viewer you can select `This Frame` -> `Open Frame in New Tab`
   - (All) In the dev tools you should be able to find the `iframe` and ctrl+click or copy the address into browser address bar
- Navigating to different pages in the pages in the object should update the URL to contain different `canvasId=` parameters  
- Reloading the page, or copying the address into a new browser window should re-open the viewer on the canvas you were previously looking at
- Manually altering the `cavasId` to be invalid, e.g. adding `foo` to the end of it, should result in the viewer loading on the first canvas
- Adding more resources and having two or more windows should mean that the first window will update the address when navigated, but the other windows should not.
   - When the page is reloaded or copied while there are multiple windows open like this, the result should be: just the initial window, showing the canvas and `canvasId` previously being viewed from the initial window.
 - Try updating the `manifestId` in the url to point to a different non-Harvard manifest, e.g. this one from Wellcome https://localhost:23017/viewer/?manifestId=https://iiif.wellcomecollection.org/presentation/v3/b13266330. The above behaviour should also apply to this manifest

# Test coverage

Yes/No: Are changes in this pull-request covered by:

- unit tests? No
- integration tests? No

# Interested parties

@enriquediaz @phil-plencner-hl
